File: Character/Services/Firebase/FirebaseImageService.swift
import Foundation
import UIKit
import FirebaseStorage

/// Firebase Storageからキャラクター画像を取得・管理するサービス
class FirebaseImageService {
    static let shared = FirebaseImageService()

    private let storage = Storage.storage()
    private let fileManager = FileManager.default
    private let cacheDirectory: URL

    // キャッシュ設定
    private let maxCacheSize: Int64 = 500 * 1024 * 1024 // 500MB
    private let cacheExpirationDays = 30

    // ダウンロード中の画像を追跡
    private var downloadTasks: [String: Task<UIImage, Error>] = [:]
    private let taskQueue = DispatchQueue(label: "com.character.firebase-image-service")

    private init() {
        // キャッシュディレクトリのセットアップ
        let cachesDirectory = fileManager.urls(for: .cachesDirectory, in: .userDomainMask)[0]
        cacheDirectory = cachesDirectory.appendingPathComponent("CharacterImages")

        // ディレクトリが存在しない場合は作成
        if !fileManager.fileExists(atPath: cacheDirectory.path) {
            try? fileManager.createDirectory(at: cacheDirectory, withIntermediateDirectories: true)
        }

        // 起動時に期限切れキャッシュをクリーン
        cleanExpiredCache()
    }

    // MARK: - Public Methods

    /// 画像を取得（キャッシュ優先）
    /// - Parameters:
    ///   - fileName: 画像ファイル名（例: "Female_HLMHL"）
    ///   - gender: 性別
    /// - Returns: UIImage
    func fetchImage(fileName: String, gender: CharacterGender) async throws -> UIImage {
        let cacheKey = "\(gender.rawValue)_\(fileName)"

        // 1. キャッシュチェック
        if let cachedImage = loadFromCache(cacheKey: cacheKey) {
            Logger.debug("🖼️ キャッシュから画像取得: \(fileName)", category: Logger.imageService)
            return cachedImage
        }

        // 2. 既にダウンロード中かチェック
        if let existingTask = taskQueue.sync(execute: { downloadTasks[cacheKey] }) {
            Logger.debug("⏳ ダウンロード中の画像を待機: \(fileName)", category: Logger.imageService)
            return try await existingTask.value
        }

        // 3. 新規ダウンロードタスクを作成
        let task = Task<UIImage, Error> {
            defer {
                taskQueue.sync {
                    downloadTasks.removeValue(forKey: cacheKey)
                }
            }

            let image = try await downloadImage(fileName: fileName, gender: gender)

            // キャッシュに保存
            saveToCache(image: image, cacheKey: cacheKey)

            return image
        }

        taskQueue.sync {
            downloadTasks[cacheKey] = task
        }

        return try await task.value
    }

    /// 画像をプリロード（バックグラウンドダウンロード）
    /// - Parameters:
    ///   - fileName: 画像ファイル名
    ///   - gender: 性別
    func preloadImage(fileName: String, gender: CharacterGender) async {
        let cacheKey = "\(gender.rawValue)_\(fileName)"

        // 既にキャッシュにある場合はスキップ
        if cacheExists(cacheKey: cacheKey) {
            return
        }

        do {
            _ = try await fetchImage(fileName: fileName, gender: gender)
            Logger.debug("✅ プリロード完了: \(fileName)", category: Logger.imageService)
        } catch {
            Logger.debug("❌ プリロード失敗: \(fileName) - \(error.localizedDescription)", category: Logger.imageService)
        }
    }

    /// キャッシュをクリア
    func clearCache() throws {
        let files = try fileManager.contentsOfDirectory(at: cacheDirectory, includingPropertiesForKeys: nil)
        for file in files {
            try fileManager.removeItem(at: file)
        }
        Logger.debug("🗑️ キャッシュをクリアしました", category: Logger.imageService)
    }

    /// キャッシュサイズを取得（バイト）
    func getCacheSize() -> Int64 {
        guard let files = try? fileManager.contentsOfDirectory(at: cacheDirectory, includingPropertiesForKeys: [.fileSizeKey]) else {
            return 0
        }

        return files.reduce(0) { total, fileURL in
            guard let resourceValues = try? fileURL.resourceValues(forKeys: [.fileSizeKey]),
                  let fileSize = resourceValues.fileSize else {
                return total
            }
            return total + Int64(fileSize)
        }
    }

    /// キャッシュサイズを人間が読める形式で取得
    func getCacheSizeFormatted() -> String {
        let bytes = getCacheSize()
        return ByteCountFormatter.string(fromByteCount: bytes, countStyle: .file)
    }

    // MARK: - Private Methods

    /// Firebase Storageから画像をダウンロード
    private func downloadImage(fileName: String, gender: CharacterGender) async throws -> UIImage {
        let storagePath = "character-images/\(gender.rawValue)/\(fileName).png"
        let storageRef = storage.reference().child(storagePath)

        Logger.debug("⬇️ Firebase Storageからダウンロード開始: \(storagePath)", category: Logger.imageService)

        // 最大サイズ: 10MB
        let maxSize: Int64 = 10 * 1024 * 1024

        do {
            let data = try await storageRef.data(maxSize: maxSize)

            guard let image = UIImage(data: data) else {
                throw FirebaseImageError.invalidImageData
            }

            Logger.debug("✅ ダウンロード完了: \(storagePath)", category: Logger.imageService)
            return image
        } catch {
            Logger.debug("❌ ダウンロード失敗: \(storagePath) - \(error.localizedDescription)", category: Logger.imageService)
            throw FirebaseImageError.downloadFailed(error)
        }
    }

    /// キャッシュから画像を読み込み
    private func loadFromCache(cacheKey: String) -> UIImage? {
        let fileURL = cacheDirectory.appendingPathComponent("\(cacheKey).png")

        guard fileManager.fileExists(atPath: fileURL.path) else {
            return nil
        }

        // ファイルの更新日時をチェック
        if let attributes = try? fileManager.attributesOfItem(atPath: fileURL.path),
           let modificationDate = attributes[.modificationDate] as? Date {
            let daysSinceModification = Calendar.current.dateComponents([.day], from: modificationDate, to: Date()).day ?? 0

            if daysSinceModification > cacheExpirationDays {
                // 期限切れのキャッシュは削除
                try? fileManager.removeItem(at: fileURL)
                return nil
            }
        }

        guard let data = try? Data(contentsOf: fileURL),
              let image = UIImage(data: data) else {
            return nil
        }

        return image
    }

    /// 画像をキャッシュに保存
    private func saveToCache(image: UIImage, cacheKey: String) {
        let fileURL = cacheDirectory.appendingPathComponent("\(cacheKey).png")

        guard let data = image.pngData() else {
            Logger.debug("❌ 画像データの変換に失敗: \(cacheKey)", category: Logger.imageService)
            return
        }

        do {
            try data.write(to: fileURL)
            Logger.debug("💾 キャッシュに保存: \(cacheKey)", category: Logger.imageService)

            // キャッシュサイズをチェックして必要に応じてクリーン
            checkCacheSizeAndClean()
        } catch {
            Logger.debug("❌ キャッシュ保存失敗: \(cacheKey) - \(error.localizedDescription)", category: Logger.imageService)
        }
    }

    /// キャッシュが存在するかチェック
    private func cacheExists(cacheKey: String) -> Bool {
        let fileURL = cacheDirectory.appendingPathComponent("\(cacheKey).png")
        return fileManager.fileExists(atPath: fileURL.path)
    }

    /// 期限切れキャッシュをクリーン
    private func cleanExpiredCache() {
        guard let files = try? fileManager.contentsOfDirectory(at: cacheDirectory, includingPropertiesForKeys: [.contentModificationDateKey]) else {
            return
        }

        let expirationDate = Calendar.current.date(byAdding: .day, value: -cacheExpirationDays, to: Date())!

        for fileURL in files {
            if let resourceValues = try? fileURL.resourceValues(forKeys: [.contentModificationDateKey]),
               let modificationDate = resourceValues.contentModificationDate,
               modificationDate < expirationDate {
                try? fileManager.removeItem(at: fileURL)
            }
        }
    }

    /// キャッシュサイズをチェックして制限を超えていたら古いファイルを削除
    private func checkCacheSizeAndClean() {
        let currentSize = getCacheSize()

        guard currentSize > maxCacheSize else {
            return
        }

        Logger.debug("⚠️ キャッシュサイズが制限を超えています: \(ByteCountFormatter.string(fromByteCount: currentSize, countStyle: .file))", category: Logger.imageService)

        // 古いファイルから削除
        guard let files = try? fileManager.contentsOfDirectory(at: cacheDirectory, includingPropertiesForKeys: [.contentModificationDateKey]) else {
            return
        }

        let sortedFiles = files.sorted { file1, file2 in
            let date1 = (try? file1.resourceValues(forKeys: [.contentModificationDateKey]))?.contentModificationDate ?? Date.distantPast
            let date2 = (try? file2.resourceValues(forKeys: [.contentModificationDateKey]))?.contentModificationDate ?? Date.distantPast
            return date1 < date2
        }

        var currentSizeAfterClean = currentSize
        for fileURL in sortedFiles {
            if currentSizeAfterClean <= maxCacheSize {
                break
            }

            if let resourceValues = try? fileURL.resourceValues(forKeys: [.fileSizeKey]),
               let fileSize = resourceValues.fileSize {
                try? fileManager.removeItem(at: fileURL)
                currentSizeAfterClean -= Int64(fileSize)
            }
        }

        Logger.debug("🗑️ キャッシュクリーン完了: \(ByteCountFormatter.string(fromByteCount: currentSizeAfterClean, countStyle: .file))", category: Logger.imageService)
    }
}

// MARK: - Error Types

enum FirebaseImageError: LocalizedError {
    case invalidImageData
    case downloadFailed(Error)
    case cacheError

    var errorDescription: String? {
        switch self {
        case .invalidImageData:
            return "画像データが無効です"
        case .downloadFailed(let error):
            return "ダウンロードに失敗しました: \(error.localizedDescription)"
        case .cacheError:
            return "キャッシュエラーが発生しました"
        }
    }
}

File: flutter/lib/data/services/firebase_image_service.dart
import 'dart:io';
import 'dart:typed_data';

import 'package:firebase_storage/firebase_storage.dart';
import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;
import 'package:path_provider/path_provider.dart';

/// キャラクター性別
enum CharacterGender {
  male('male'),
  female('female');

  final String value;
  const CharacterGender(this.value);
}

/// Firebase Storageからキャラクター画像を取得・管理するサービス
class FirebaseImageService {
  static final FirebaseImageService shared = FirebaseImageService._();

  final FirebaseStorage _storage = FirebaseStorage.instance;
  Directory? _cacheDirectory;
  bool _isInitialized = false;

  // キャッシュ設定
  static const int _maxCacheSize = 500 * 1024 * 1024; // 500MB
  static const int _cacheExpirationDays = 30;

  // ダウンロード中の画像を追跡
  final Map<String, Future<Uint8List>> _downloadTasks = {};

  // メモリキャッシュ（Web用）
  final Map<String, Uint8List> _memoryCache = {};

  // URLキャッシュ（Firebase Storage APIへのリクエストを削減）
  final Map<String, String> _urlCache = {};

  FirebaseImageService._();

  /// 初期化
  Future<void> initialize() async {
    if (_isInitialized) return;

    // Webではpath_providerが使えないのでメモリキャッシュのみ
    if (!kIsWeb) {
      try {
        final cacheDir = await getApplicationCacheDirectory();
        _cacheDirectory = Directory('${cacheDir.path}/CharacterImages');

        if (!await _cacheDirectory!.exists()) {
          await _cacheDirectory!.create(recursive: true);
        }

        // 起動時に期限切れキャッシュをクリーン
        await _cleanExpiredCache();
      } catch (e) {
        debugPrint('Failed to initialize file cache: $e');
      }
    }

    _isInitialized = true;
  }

  /// 画像を取得（キャッシュ優先）
  Future<Uint8List> fetchImage({
    required String fileName,
    required CharacterGender gender,
  }) async {
    // 初期化されていない場合は初期化
    if (!_isInitialized) {
      await initialize();
    }

    final cacheKey = '${gender.value}_$fileName';

    // 1. メモリキャッシュチェック（Web/モバイル共通）
    if (_memoryCache.containsKey(cacheKey)) {
      debugPrint('🖼️ メモリキャッシュから画像取得: $fileName');
      return _memoryCache[cacheKey]!;
    }

    // 2. ファイルキャッシュチェック（モバイルのみ）
    if (!kIsWeb) {
      final cachedImage = await _loadFromCache(cacheKey);
      if (cachedImage != null) {
        debugPrint('🖼️ ファイルキャッシュから画像取得: $fileName');
        _memoryCache[cacheKey] = cachedImage;
        return cachedImage;
      }
    }

    // 3. 既にダウンロード中かチェック
    if (_downloadTasks.containsKey(cacheKey)) {
      debugPrint('⏳ ダウンロード中の画像を待機: $fileName');
      return _downloadTasks[cacheKey]!;
    }

    // 4. 新規ダウンロードタスクを作成
    final task = _downloadImage(fileName: fileName, gender: gender)
        .then((data) async {
      // メモリキャッシュに保存
      _memoryCache[cacheKey] = data;

      // ファイルキャッシュに保存（モバイルのみ）
      if (!kIsWeb) {
        await _saveToCache(data, cacheKey);
      }

      _downloadTasks.remove(cacheKey);
      return data;
    }).catchError((error) {
      _downloadTasks.remove(cacheKey);
      throw error;
    });

    _downloadTasks[cacheKey] = task;
    return task;
  }

  /// 画像をプリロード（バックグラウンドダウンロード）
  Future<void> preloadImage({
    required String fileName,
    required CharacterGender gender,
  }) async {
    final cacheKey = '${gender.value}_$fileName';

    // 既にキャッシュにある場合はスキップ
    if (await _cacheExists(cacheKey)) {
      return;
    }

    try {
      await fetchImage(fileName: fileName, gender: gender);
      debugPrint('✅ プリロード完了: $fileName');
    } catch (e) {
      debugPrint('❌ プリロード失敗: $fileName - $e');
    }
  }

  /// 画像URLを取得（URLキャッシュ付き）
  Future<String> getImageUrl({
    required String fileName,
    required CharacterGender gender,
  }) async {
    final cacheKey = '${gender.value}_$fileName';
    if (_urlCache.containsKey(cacheKey)) {
      return _urlCache[cacheKey]!;
    }
    final storagePath = 'character-images/${gender.value}/$fileName.png';
    final ref = _storage.ref().child(storagePath);
    final url = await ref.getDownloadURL();
    _urlCache[cacheKey] = url;
    return url;
  }

  /// キャッシュをクリア
  Future<void> clearCache() async {
    if (_cacheDirectory == null) return;

    final files = await _cacheDirectory!.list().toList();
    for (final file in files) {
      await file.delete();
    }
    debugPrint('🗑️ キャッシュをクリアしました');
  }

  /// キャッシュサイズを取得（バイト）
  Future<int> getCacheSize() async {
    if (_cacheDirectory == null) return 0;

    int totalSize = 0;
    final files = await _cacheDirectory!.list().toList();
    for (final file in files) {
      if (file is File) {
        totalSize += await file.length();
      }
    }
    return totalSize;
  }

  /// キャッシュサイズを人間が読める形式で取得
  Future<String> getCacheSizeFormatted() async {
    final bytes = await getCacheSize();
    if (bytes < 1024) return '$bytes B';
    if (bytes < 1024 * 1024) return '${(bytes / 1024).toStringAsFixed(1)} KB';
    return '${(bytes / (1024 * 1024)).toStringAsFixed(1)} MB';
  }

  /// Firebase Storageから画像をダウンロード
  Future<Uint8List> _downloadImage({
    required String fileName,
    required CharacterGender gender,
  }) async {
    final storagePath = 'character-images/${gender.value}/$fileName.png';
    final ref = _storage.ref().child(storagePath);

    debugPrint('⬇️ Firebase Storageからダウンロード開始: $storagePath');

    try {
      // Webの場合はgetDownloadURLを使用してHTTP経由でダウンロード
      if (kIsWeb) {
        final url = await ref.getDownloadURL();
        debugPrint('📎 ダウンロードURL取得: $url');

        // HTTP経由で画像をダウンロード
        final response = await _httpGet(url);
        if (response != null) {
          debugPrint('✅ ダウンロード完了: $storagePath');
          return response;
        }
        throw FirebaseImageException.downloadFailed('HTTP download failed');
      } else {
        // モバイルの場合は直接getData
        const maxSize = 10 * 1024 * 1024;
        final data = await ref.getData(maxSize);

        if (data == null) {
          throw FirebaseImageException.invalidImageData();
        }

        debugPrint('✅ ダウンロード完了: $storagePath');
        return data;
      }
    } catch (e) {
      debugPrint('❌ ダウンロード失敗: $storagePath - $e');
      throw FirebaseImageException.downloadFailed(e.toString());
    }
  }

  /// HTTP経由で画像をダウンロード（Web用）
  Future<Uint8List?> _httpGet(String url) async {
    try {
      final response = await http.get(Uri.parse(url));
      if (response.statusCode == 200) {
        return response.bodyBytes;
      }
      debugPrint('HTTP download failed: ${response.statusCode}');
    } catch (e) {
      debugPrint('HTTP download error: $e');
    }
    return null;
  }

  /// キャッシュから画像を読み込み
  Future<Uint8List?> _loadFromCache(String cacheKey) async {
    if (_cacheDirectory == null) return null;

    final file = File('${_cacheDirectory!.path}/$cacheKey.png');

    if (!await file.exists()) {
      return null;
    }

    // ファイルの更新日時をチェック
    final stat = await file.stat();
    final daysSinceModification =
        DateTime.now().difference(stat.modified).inDays;

    if (daysSinceModification > _cacheExpirationDays) {
      // 期限切れのキャッシュは削除
      await file.delete();
      return null;
    }

    return await file.readAsBytes();
  }

  /// 画像をキャッシュに保存
  Future<void> _saveToCache(Uint8List data, String cacheKey) async {
    if (_cacheDirectory == null) return;

    final file = File('${_cacheDirectory!.path}/$cacheKey.png');

    try {
      await file.writeAsBytes(data);
      debugPrint('💾 キャッシュに保存: $cacheKey');

      // キャッシュサイズをチェックして必要に応じてクリーン
      await _checkCacheSizeAndClean();
    } catch (e) {
      debugPrint('❌ キャッシュ保存失敗: $cacheKey - $e');
    }
  }

  /// キャッシュが存在するかチェック
  Future<bool> _cacheExists(String cacheKey) async {
    if (_cacheDirectory == null) return false;

    final file = File('${_cacheDirectory!.path}/$cacheKey.png');
    return await file.exists();
  }

  /// 期限切れキャッシュをクリーン
  Future<void> _cleanExpiredCache() async {
    if (_cacheDirectory == null) return;

    final expirationDate =
        DateTime.now().subtract(const Duration(days: _cacheExpirationDays));

    final files = await _cacheDirectory!.list().toList();
    for (final file in files) {
      if (file is File) {
        final stat = await file.stat();
        if (stat.modified.isBefore(expirationDate)) {
          await file.delete();
        }
      }
    }
  }

  /// キャッシュサイズをチェックして制限を超えていたら古いファイルを削除
  Future<void> _checkCacheSizeAndClean() async {
    final currentSize = await getCacheSize();

    if (currentSize <= _maxCacheSize) {
      return;
    }

    debugPrint('⚠️ キャッシュサイズが制限を超えています: ${await getCacheSizeFormatted()}');

    if (_cacheDirectory == null) return;

    // 古いファイルから削除
    final files = await _cacheDirectory!.list().toList();
    final fileStats = <FileSystemEntity, FileStat>{};

    for (final file in files) {
      if (file is File) {
        fileStats[file] = await file.stat();
      }
    }

    final sortedFiles = files
      ..sort((a, b) {
        final statA = fileStats[a];
        final statB = fileStats[b];
        if (statA == null || statB == null) return 0;
        return statA.modified.compareTo(statB.modified);
      });

    int currentSizeAfterClean = currentSize;
    for (final file in sortedFiles) {
      if (currentSizeAfterClean <= _maxCacheSize) {
        break;
      }

      if (file is File) {
        final size = await file.length();
        await file.delete();
        currentSizeAfterClean -= size;
      }
    }

    debugPrint('🗑️ キャッシュクリーン完了');
  }
}

/// Firebase画像エラー
class FirebaseImageException implements Exception {
  final String message;

  FirebaseImageException._(this.message);

  factory FirebaseImageException.invalidImageData() {
    return FirebaseImageException._('画像データが無効です');
  }

  factory FirebaseImageException.downloadFailed(String error) {
    return FirebaseImageException._('ダウンロードに失敗しました: $error');
  }

  factory FirebaseImageException.cacheError() {
    return FirebaseImageException._('キャッシュエラーが発生しました');
  }

  @override
  String toString() => message;
}
